pkg/pgmodel/migrate.go

```diff
@@ -26,133 +28,344 @@ const (
 	extensionInstall            = "CREATE EXTENSION IF NOT EXISTS timescale_prometheus_extra WITH SCHEMA %s;"
 	metadataUpdateWithExtension = "SELECT update_tsprom_metadata($1, $2, $3)"
 	metadataUpdateNoExtension   = "INSERT INTO _timescaledb_catalog.metadata(key, value, include_in_telemetry) VALUES ($1, $2, $3) ON CONFLICT (key) DO UPDATE SET value = EXCLUDED.value, include_in_telemetry = EXCLUDED.include_in_telemetry"
+	createMigrationsTable       = "CREATE TABLE IF NOT EXISTS prom_schema_migrations (version text not null primary key)"
+	getVersion                  = "SELECT version FROM prom_schema_migrations LIMIT 1"
+	setVersion                  = "INSERT INTO prom_schema_migrations (version) VALUES ($1)"
+	truncateMigrationsTable     = "TRUNCATE prom_schema_migrations"
+
+	preinstallScripts = "preinstall"
+	versionScripts    = "versions"
+	idempotentScripts = "idempotent"
 )
 
-var ExtensionIsInstalled = false
+var (
+	ExtensionIsInstalled = false
 
-type mySrc struct {
-	source.Driver
-}
+	toc = map[string][]string{
+		"idempotent": {},
+	}
+	migrateMutex = &sync.Mutex{}
+)
 
 type VersionInfo struct {
 	Version    string
 	CommitHash string
 }
 
-func (t *mySrc) replaceSchemaNames(r io.ReadCloser) (io.ReadCloser, error) {
-	buf := new(bytes.Buffer)
-	_, err := buf.ReadFrom(r)
+type prefixedName struct {
+	prefix int
+	name   string
+}
+
+type prefixedNames []prefixedName
+
+func (p prefixedNames) Len() int {
+	return len(p)
+}
+
+func (p prefixedNames) Less(i, j int) bool {
+	return p[i].prefix < p[j].prefix
+}
+
+func (p prefixedNames) Swap(i, j int) {
+	p[i], p[j] = p[j], p[i]
+}
+
+func (p prefixedNames) getNames() []string {
+	names := make([]string, len(p))
+	for i, e := range p {
+		names[i] = e.name
+	}
+	return names
+}
+
+// Migrate performs a database migration to the latest version
+func Migrate(db *pgxpool.Pool, versionInfo VersionInfo) (err error) {
+	migrateMutex.Lock()
+	defer migrateMutex.Unlock()
+
+	// Getting an early connection to install the extra extension.
+	// TODO: Investigate why this is required. Installing the extension on the
+	// migration connection fails, thats why this is necessary. This should be removed.
+	conn, err := db.Acquire(context.Background())
 	if err != nil {
-		return r, err
+		return fmt.Errorf("could not acquire connection: %w", err)
 	}
-	err = r.Close()
+
+	defer conn.Release()
+
+	appVersion, err := semver.Make(versionInfo.Version)
 	if err != nil {
-		return r, err
+		return fmt.Errorf("app version is not semver format, aborting migration")
+	}
+	if err := ensureVersionTable(db); err != nil {
+		return fmt.Errorf("error ensuring version table: %w", err)
+	}
+	dbVersion, err := getDBVersion(db)
+	if err != nil {
+		return fmt.Errorf("failed to get the version from database: %w", err)
+	}
```

Comment:
I suspect we'll eventually need a way to handle a malformed db-version, probably by forcing it to assume the db has a specific version, but I'm fine leaving that for a latter date